### PR TITLE
Feature/cxec 9980 Integrate APIs in create new ticket dialog

### PR DESCRIPTION
**Pull request by manpreetlamba4**
This PR contains changes for adding 'Categories' API and 'Associated Objects' API in the customer ticketing create dialog

https://jira.tools.sap/browse/CXEC-16778
https://jira.tools.sap/browse/CXEC-16779

### CODE REVIEW

**Review by adolfobee — APPROVED** [review]

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html [hunk 1]
```diff
@@ -52,7 +54,12 @@
             {{ 'customerTicketing.associateTo' | cxTranslate }}
           </span>
           <select class="form-control">
-            <option></option>
+            <option
+              *ngFor="let objects of ticketAssociatedObjects"
```
Comment (khristian23) [inline feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html]:
```suggestion
              *ngFor="let associatedObject of ticketAssociatedObjects"
```
should be in singular

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html [hunk 2]
```diff
@@ -52,7 +54,12 @@
...
+              [ngValue]="c"
```
Comment (khristian23) [inline feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html]: What does this mean ngValue = c?

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts [hunk 3]
```diff
@@ -10,7 +13,18 @@ export class CustomerTicketingCreateDialogComponent
   extends CustomerTicketingDialogComponent
   implements OnInit
 {
+  ticketCategories: Array<Category>;
+  ticketAssociatedObjects: Array<AssociatedObjects>;
+
   ngOnInit(): void {
+    this.customerTicketingFacade.getTicketCategories().subscribe((category) => {
```
Comment (khristian23) [inline feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts]: shouldn't it be categories in plural?

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts [hunk 4]
```diff
@@ -10,7 +13,18 @@ export class CustomerTicketingCreateDialogComponent
...
+      this.ticketCategories = category;
+    });
+    this.customerTicketingFacade
+      .getTicketAssociatedObjects()
+      .subscribe((associatedObject) => {
```
Comment (khristian23) [inline feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts]: Same here, it should be plural

File: feature-libs/customer-ticketing/root/model/customer-ticketing.model.ts [hunk 5]
```diff
@@ -29,6 +29,20 @@ export interface Category {
   name: string;
 }
 
+export interface CategoryList {
+  ticketCategories?: Category[];
+}
+
+export interface AssociatedObjects {
```
Comment (khristian23) [inline feature-libs/customer-ticketing/root/model/customer-ticketing.model.ts]:
```suggestion
export interface AssociatedObject {
```
This interface represents only one Object.  It should be in singular

File: feature-libs/customer-ticketing/root/model/customer-ticketing.model.ts [hunk 6]
```diff
@@ -29,6 +29,20 @@ export interface Category {
   name: string;
 }
 
+export interface CategoryList {
```
Comment (khristian23) [inline feature-libs/customer-ticketing/root/model/customer-ticketing.model.ts]:
```suggestion
export interface CategoriesList {
```
Or ListOfCategories
Or categoryListing

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts [hunk 7]
```diff
@@ -19,6 +25,18 @@ import { CustomerTicketingConnector } from '../connectors';
...
+  }
+  getTicketCategoriesQueryResetEvents(): QueryNotifier[] | undefined {
+    return [GetTicketCatQueryResetEvent];
+  }
+  getTicketAssociatedObjectsQueryReloadEvents(): QueryNotifier[] | undefined {
+    return [GetTicketAssociatedObjectsQueryReloadEvent];
+  }
+  getTicketAssociatedObjectsQueryResetEvents(): QueryNotifier[] | undefined {
+    return [GetTicketAssociatedObjectsQueryResetEvent];
+  }
```
Comment (khristian23) [inline feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts]: Why do these methods return. QueryNotifier[] or undefined?
Shouldn't be enough to return QueryNotifier[], since their implementations are hardcoded to return an array of one element?

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts [hunk 8]
```diff
@@ -70,6 +130,19 @@ export class CustomerTicketingService implements CustomerTicketingFacade {
       })
     );
   }
+  protected customerTicketingAssociatedObjectsPreConditions(): Observable<string> {
+    return this.userIdService.getUserId().pipe(
+      take(1),
+      map((userId) => {
+        if (!userId) {
+          throw new Error(
+            'Customer ticketing associated objects pre conditions not met'
```
Comment (khristian23) [inline feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts:139]: Is this a message that can be seen by the customer?  If so, you need to make it transalatable

File: feature-libs/customer-ticketing/occ/adapters/occ-customer-ticketing.adapter.ts [hunk 9]
```diff
@@ -20,6 +28,44 @@ export class OccCustomerTicketingAdapter implements CustomerTicketingAdapter {
...
+      },
+    });
+  }
+
+  getTicketCategories(): Observable<Category[]> {
+    return this.http.get<CategoryList>(this.getTicketCategoriesEndpoint()).pipe(
+      catchError((error) => throwError(normalizeHttpError(error))),
+      map((categoryList) => categoryList.ticketCategories ?? []),
+      this.converter.pipeableMany(CUSTOMER_TICKETING_CATEGORY_NORMALIZER)
+    );
```
Comment (khristian23) [inline feature-libs/customer-ticketing/occ/adapters/occ-customer-ticketing.adapter.ts]: Cna it be pretty printed?

**Review by khristian23 — CHANGES_REQUESTED** [review]
I understand changing interface AssociatedObjects by its correspondent singular AssociatedObject will imply changing many places, but it should not be plural since it represents a single object

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html (hunk 2 above)
Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html]: I'd like to know to know as well  +1 


My current assumption is that it is dead code as this form is not doing anything. 'c' was supposed to be objects, I believe as well

First, you would need to attach an event listener, specifically the change event (`select (change)='etcFunctionToupdateTheForm(passingTheValueSelected)'`) to the select tag, which you would create in order to update the form

---
If you search for `<option` in the codebase, you can see how to attach an event listener, and what to do with it

Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html]: Just to add more to Christian's other comments, it would have to be singular, similarly to how you named things in line 44 of this fine.

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html [hunk 10]
```diff
@@ -41,7 +41,9 @@
             {{ 'customerTicketing.category' | cxTranslate }}
           </span>
           <select class="form-control">
-            <option></option>
+            <option *ngFor="let category of ticketCategories">
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.html]: similarly to https://github.com/SAP/spartacus/pull/16142/files#r988362362

it doesn't do anything. It just 'displays' a dropdown, but the values aren't being passed to the APIs

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts (hunk 3 above)
Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts]: If you were to subscribe in the component .ts file, you would need to always unsubscribe on component destroy as it will cause memory leaks.

You have options:

1 - let angular handle everything by using the `async` pipe in the template ('.html'). What this does for you is it subscribes to the observable, and unsubscribes to it whenever the component gets destroyed.


2 - Do what you're doing here, but you would need to create a subscription object, such as `const subscriptions = new Subscription()` from `rxjs`. and add every subscribed things in the component level. You would need to implement the `OnDestroy` to handle the lifecycle when the component gets destryoed, and therefore unsubscribe in one shot.

Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts]: I would prefer, you would go with the 1st approach as it's less code and more compact. 

That would also mean, you would make the variables of type Observable< insert type here >

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts (hunk 4 above)
Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts]: same subscription problem here https://github.com/SAP/spartacus/pull/16142/files#r988362872

File: feature-libs/customer-ticketing/components/shared/customer-ticketing-dialog/customer-ticketing-dialog.component.spec.ts [hunk 11]
```diff
@@ -13,6 +14,7 @@ describe('CustomerTicketingDialogComponent', () => {
 
   beforeEach(async () => {
     await TestBed.configureTestingModule({
+      imports: [I18nTestingModule],
       declarations: [CustomerTicketingDialogComponent],
     }).compileComponents();
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/components/shared/customer-ticketing-dialog/customer-ticketing-dialog.component.spec.ts:19]: Is there a dedicated ticket for unit tests here? I thought it would be done once the apis will be integrated in conjunction with the template that was created a while back

File: feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts [hunk 12]
```diff
@@ -10,7 +13,18 @@ export class CustomerTicketingCreateDialogComponent
...
+        this.ticketAssociatedObjects = associatedObject;
+      });
     this.buildForm();
   }
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/components/customer-ticketing-create/customer-ticketing-create-dialog/customer-ticketing-create-dialog.component.ts]: createTicketRequest  --- line 17

you would need to utilize this function to pass to the newly integrated service to use your apis

File: feature-libs/customer-ticketing/components/shared/customer-ticketing-dialog/customer-ticketing-dialog.component.spec.ts (hunk 11 above)
Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/components/shared/customer-ticketing-dialog/customer-ticketing-dialog.component.spec.ts:19]: Applies to all specs here so I don't flag every single one of them

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.spec.ts [hunk 13]
```diff
@@ -93,4 +93,62 @@ describe('CustomerTicketingService', () => {
         });
     });
   });
+  describe('getTicketCategories', () => {
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/core/facade/customer-ticketing.service.spec.ts:96]: nice 👍

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts (hunk 8 above)
Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts:139]: You can re-use the `customerTicketingPreConditions` as it returns a tuple.

The current logic says if ticketId doesn't exist, throw an error. We can just remove that condition and make it able to return a tuple of [userId, ticketId || undefined]. The way to do this is to make sure your second stream that is being combined can startWith `undefined`

Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts:139]: Because the way I see it, we don't always need the ticketId. Only for certain cases when it's within the ticket details page.

File: feature-libs/customer-ticketing/root/events/customer-ticketing-event.listener.ts [hunk 14]
```diff
@@ -45,6 +49,56 @@ export class CustomerTicketingEventListener implements OnDestroy {
     );
   }
 
+  protected onGetTicketCatQueryReset(): void {
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/root/events/customer-ticketing-event.listener.ts]: 1- for all these created functions, they needed to be called in the constructor

2 - we don't need all these functions, you can place this in one function that has the merges of the stream of `LogoutEvent` and `LoginEvent`, which will fire those events all in one because at this point, it is just DRY

-----
The 1- would make sense if we're making multiple event listeners for each of the `facade`, but it looks like the direction we want to go is to have one facade that encapsulates the whole functionality. I don't have much preference. It is entirely up to you guys on whether you want to have decouple the facade into separate ones to keep the service logic different. If not, then let's go with 2.

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]
1 - missing a lot of functionality in terms of integrating the apis for creating a new ticket.
- the component and apis aren't fully integrated. User can't create a ticket, nor the api to create the ticket is done

2 - dialog only contains to display data, but not working with it, for example, updating the form based on what is selected and doing something with it

**Comment by bgambocjaviniar** [comment]
> 1 - missing a lot of functionality in terms of integrating the apis for creating a new ticket.
> 
> * the component and apis aren't fully integrated. User can't create a ticket, nor the api to create the ticket is done
> 
> 2 - dialog only contains to display data, but not working with it, for example, updating the form based on what is selected and doing something with it

scratch 1-

The title misled me to believe this PR is supposed to integrate the entire API for creating a dialog, but apparently it's just to get the categories and associated objects in the dropdown. But, the 2- still applies, where the form is not going to be updated as it's not fully integrated with api.

I am confused though as the branch name is referencing this jira, which would mean it 'should' include all the apis and fully integrated with your template. https://jira.tools.sap/browse/CXEC-9980.

I would create the appropriate branch that corresponds to what this PR is actually for and change the title of the PR

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts [hunk 15]
```diff
@@ -45,13 +63,55 @@ export class CustomerTicketingService implements CustomerTicketingFacade {
...
+    QueryState<AssociatedObject[]>
+  > {
+    return this.getTicketAssociatedObjectsQuery.getState();
+  }
+  getTicketAssociatedObjects(): Observable<AssociatedObject[]> {
+    return this.getTicketAssociatedObjectsState().pipe(
+      map((state) => state.data ?? [])
+    );
+  }
+  getTicketCategories(): Observable<Category[]> {
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts:106]: Let's switch the function just for the sake of reading the 'service', where it's first the state then the 'data'.

Basically swap line 106 function with line 1123 function

```
  getTicketCategoriesState(): Observable<QueryState<Category[]>> {
    return this.getTicketCategoriesQuery.getState();
  }

getTicketCategories(): Observable<Category[]> {
....
}
```

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts [hunk 16]
```diff
@@ -63,13 +123,26 @@ export class CustomerTicketingService implements CustomerTicketingFacade {
...
       map(([userId, ticketId]) => {
-        if (!userId || !ticketId) {
+        if (!userId) {
           throw new Error('Customer ticketing pre conditions not met');
         }
         return [userId, ticketId];
       })
     );
   }
+  protected customerTicketingAssociatedObjectsPreConditions(): Observable<string> {
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts:133]: needs to be removed as it's not longer used

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts [hunk 17]
```diff
@@ -63,13 +123,26 @@ export class CustomerTicketingService implements CustomerTicketingFacade {
...
+      map((userId) => {
+        if (!userId) {
+          throw new Error(
+            'Customer ticketing associated objects pre conditions not met'
+          );
+        }
+        return userId;
+      })
+    );
+  }
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/core/facade/customer-ticketing.service.ts:145]: nit, can you move getTicketState and getTicket above the precondition function

File: feature-libs/customer-ticketing/root/events/customer-ticketing.events.ts [hunk 18]
```diff
@@ -7,3 +7,19 @@ export class GetTicketQueryResetEvent extends CxEvent {
 export class GetTicketQueryReloadEvent extends CxEvent {
   static readonly type = 'GetTicketQueryReloadEvent';
 }
+
+export class GetTicketCatQueryResetEvent extends CxEvent {
```
Comment (bgambocjaviniar) [inline feature-libs/customer-ticketing/root/events/customer-ticketing.events.ts:11]: Let's explicitly say Category instead of Cat

Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/root/events/customer-ticketing.events.ts:11]: same for other instances in this PR

File: feature-libs/customer-ticketing/core/facade/customer-ticketing.service.spec.ts (hunk 13 above)
Comment (bgambocjaviniar) [reply feature-libs/customer-ticketing/core/facade/customer-ticketing.service.spec.ts:96]: These tests are actually failing

have you ran `ng test customer-tickerting`. I said nice above because I saw tests, but they're not passing. 
What needs to be done is that it requires the mock connector to be aware of these functions, so include them here https://github.com/SAP/spartacus/pull/16142/files#diff-450d65a6cbacd13bee796c96a548940c8ee202110c969f6fe9e25a335fda26ddR32

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]
1 - first thing first. Let's fix the convention on creating branches that relates to a jira ticket. Let's change this head branch to feature/CXEC-16778 instead (which would include changes from 16779, which is fine), but it's part of a subtask. If we were to merge this branch, it's like saying  you fully integrated the API to create a new ticket dialog, but that is false as 9980 is not done. I mentioned here to create a new branch and change the title https://github.com/SAP/spartacus/pull/16142#issuecomment-1275549610

2 - I un-resolved this https://github.com/SAP/spartacus/pull/16142#discussion_r992958938. What I want is a function that handles firing all the expected reset/reload on login / reload. 

Basically, something like below

```
onLoginAndLogoutEvent(): {
this.subscriptions.add(
      merge(
        this.eventService.get(LogoutEvent),
        this.eventService.get(LoginEvent)
      ).subscribe(() => {
dispatch all reload / reset of everything that hooks into login/logout in this file to avoid DRY. <---- important
}
}
```

3 - the service unit tests aren't passing, look at the comments I wrote on how to fix. I also left other tiny fixes to do

**Comment by manpreetlamba4** [comment]
All comments are resolved in https://github.com/SAP/spartacus/pull/16301
